### PR TITLE
CORE-6668: Simulator service overrides

### PR DESCRIPTION
**Pull request by LizR3**
Simulator configuration allows Corda services to be overridden or decorated (with the original service from Simulator passed into the custom service).

### CODE REVIEW

File: simulator/runtime/src/test/kotlin/net/corda/simulator/runtime/flows/DefaultServicesInjectorTest.kt [hunk 1]
```diff
@@ -33,4 +45,77 @@ class DefaultServicesInjectorTest {
...
+        val fiber = SimFiberBase()
+
+        fiber.use {
+            // When we inject services into it
+            DefaultServicesInjector(config).injectServices(flow, member, it)
+        }
+
+        // Then the flow should have the overridden services
+        assertThat(flow.jsonMarshallingService, `is`(myJsonMarshallingService))
+        assertThat(flow.merkleTreeFactory, `is`(myFactory))
```
Comment (mattbradburyr3) [inline simulator/runtime/src/test/kotlin/net/corda/simulator/runtime/flows/DefaultServicesInjectorTest.kt:103]: Do we want to test that other services which are injected that haven't been overriden still get the default service injected - ie the override mechanism for one service hasn't interfered with the other services's injections

File: simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/utils/Utilities.kt [hunk 2]
```diff
@@ -18,38 +19,41 @@ import net.corda.v5.base.types.MemberX500Name
  * @param fieldClass The type of the field.
  * @param valueCreator A factory method to create the value to assign to the field.
  */
-fun <T> Flow.injectIfRequired(
-    fieldClass: Class<T>,
-    valueCreator: () -> T
+fun Flow.injectIfRequired(
+    fieldClass: Class<*>,
+    valueCreator: () -> Any
```
Comment (mattbradburyr3) [inline simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/utils/Utilities.kt:24]: Why are we changing this from T to */Any

**Review by mattbradburyr3 — COMMENTED** [review]
I've traced the code through and it makes sense + checked through the tests.

No blockers but a few clarifying questions.

File: simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/utils/Utilities.kt (hunk 2 above)
Comment (LizR3) [reply simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/utils/Utilities.kt:24]: The class type isn't actually needed for the injection, and the overrides are (by virtue of being in a mixed list) unable to store their type. This lets us pass the fields provided by the override builders into the same injection mechanism as other builders.

File: simulator/runtime/src/test/kotlin/net/corda/simulator/runtime/flows/DefaultServicesInjectorTest.kt (hunk 1 above)
Comment (LizR3) [reply simulator/runtime/src/test/kotlin/net/corda/simulator/runtime/flows/DefaultServicesInjectorTest.kt:103]: Sure; it's a small enough change.

File: simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/utils/Utilities.kt (hunk 2 above)
Comment (LizR3) [reply simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/utils/Utilities.kt:24]: (Note that the signature of the ServiceOverrideBuilder as well as the SimulatorConfigurationBuilder API prevent anyone from adding an override that doesn't match the class.)

**Review by mattbradburyr3 — APPROVED** [review]

**Pull request merged**